### PR TITLE
openapi3filter: validate non-string headers

### PR DESCRIPTION
**Pull request by slessard**
Previously the code assumed response headers were only strings. Now the code can validate any type of response header

### CODE REVIEW

File: openapi3/parameter.go [hunk 1]
```diff
@@ -90,7 +90,7 @@ func (parameters Parameters) Validate(ctx context.Context, opts ...ValidationOpt
 }
 
 // Parameter is specified by OpenAPI/Swagger 3.0 standard.
-// See https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#parameterObject
+// See https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#parameter-object
```
Comment (fenollp) [inline openapi3/parameter.go:93]: Thanks for this but please revert, I'll fix them all in another PR.

**Review by fenollp — CHANGES_REQUESTED** [review]
Thanks for this. Please add code testing the feature you're adding.

File: openapi3/parameter.go (hunk 1 above)
Comment (slessard) [reply openapi3/parameter.go]: I've reverted this change

**Review by fenollp — CHANGES_REQUESTED** [review]
Please add a test we're good :)

**Comment by slessard** [comment]
Unit tests have been added. In the process I found and fixed a few related bugs

File: openapi3filter/validate_response.go [hunk 2]
```diff
@@ -171,6 +158,38 @@ func ValidateResponse(ctx context.Context, input *ResponseValidationInput) error
 	return nil
 }
 
+func validateResponseHeader(headerName string, headerRef *openapi3.HeaderRef, input *ResponseValidationInput, opts []openapi3.SchemaValidationOption, ) error {
+	var err error
+	var decodedValue interface{}
+	var found bool
+	dec := &headerParamDecoder{header: input.Header}
+	sm, _ := headerRef.Value.SerializationMethod() // ignore error return because a header's serialization method getter does not return an error
```
Comment (fenollp) [inline openapi3filter/validate_response.go:166]: Please do check the error.

File: openapi3filter/validate_response.go [hunk 3]
```diff
@@ -78,24 +78,11 @@ func ValidateResponse(ctx context.Context, input *ResponseValidationInput) error
...
-		if err := s.Value.Schema.Value.VisitJSON(h, opts...); err != nil {
-			return &ResponseError{
-				Input:  input,
-				Reason: fmt.Sprintf("response header %q doesn't match the schema", k),
-				Err:    err,
-			}
+	for _, headerName := range headers {
+		headerRef := response.Headers[headerName]
+		err := validateResponseHeader(headerName, headerRef, input, opts)
+		if err != nil {
```
Comment (fenollp) [inline openapi3filter/validate_response.go:84]:
```suggestion
		if err := validateResponseHeader(headerName, headerRef, input, opts); err != nil {
```

**Review by fenollp — CHANGES_REQUESTED** [review]
Please the formatter with this diff:
```diff
diff --git a/openapi3filter/validate_response.go b/openapi3filter/validate_response.go
index 1108031..84669b0 100644
--- a/openapi3filter/validate_response.go
+++ b/openapi3filter/validate_response.go
@@ -158,7 +158,7 @@ func ValidateResponse(ctx context.Context, input *ResponseValidationInput) error
        return nil
 }
 
-func validateResponseHeader(headerName string, headerRef *openapi3.HeaderRef, input *ResponseValidationInput, opts []openapi3.SchemaValidationOption, ) error {
+func validateResponseHeader(headerName string, headerRef *openapi3.HeaderRef, input *ResponseValidationInput, opts []openapi3.SchemaValidationOption) error {
        var err error
        var decodedValue interface{}
        var found bool
diff --git a/openapi3filter/validate_response_test.go b/openapi3filter/validate_response_test.go
index 71e1062..5ce657b 100644
--- a/openapi3filter/validate_response_test.go
+++ b/openapi3filter/validate_response_test.go
@@ -6,8 +6,9 @@ import (
        "strings"
        "testing"
 
-       "github.com/getkin/kin-openapi/openapi3"
        "github.com/stretchr/testify/require"
+
+       "github.com/getkin/kin-openapi/openapi3"
 )
 
 func Test_validateResponseHeader(t *testing.T) {
@@ -211,4 +212,4 @@ func newArraySchema(schema *openapi3.Schema) *openapi3.Schema {
        arraySchema.Items = openapi3.NewSchemaRef("", schema)
 
        return arraySchema
-}
\ No newline at end of file
+}
```
but then I see test failures:
```
FAIL | Issue201 (0.00s)
FAIL |   Issue201/missing_required_header (0.00s)
     |         issue201_test.go:132: 
     |             	Error Trace:	/home/pete/Documents/_cake/kin-openapi.git/openapi3filter/issue201_test.go:132
     |             	Error:      	Error "response header \"X-Blap\" doesn't match schema: Value is not nullable\nSchema:\n  {\n    \"pattern\": \"^blap$\"\n  }\n\nValue:\n  null\n" does not contain "response header \"X-Blip\" missing"
     |             	Test:       	TestIssue201/missing_required_header
FAIL |   Issue201/invalid_required_header (0.00s)
     |         issue201_test.go:132: 
     |             	Error Trace:	/home/pete/Documents/_cake/kin-openapi.git/openapi3filter/issue201_test.go:132
     |             	Error:      	Error "response header \"X-Blap\" doesn't match schema: Value is not nullable\nSchema:\n  {\n    \"pattern\": \"^blap$\"\n  }\n\nValue:\n  null\n" does not contain "response header \"X-Blup\" doesn't match schema: string \"bluuuuuup\" doesn't match the regular expression \"^blup$\""
     |             	Test:       	TestIssue201/invalid_required_header
FAIL |   Issue201/no_error (0.00s)
     |         issue201_test.go:134: 
     |             	Error Trace:	/home/pete/Documents/_cake/kin-openapi.git/openapi3filter/issue201_test.go:134
     |             	Error:      	Received unexpected error:
     |             	            	response header "X-Blap" doesn't match schema: Value is not nullable
     |             	            	Schema:
     |             	            	  {
     |             	            	    "pattern": "^blap$"
     |             	            	  }
     |             	            	
     |             	            	Value:
     |             	            	  null
     |             	Test:       	TestIssue201/no_error
FAIL |   Issue201/missing_non-required_header (0.00s)
     |         issue201_test.go:134: 
     |             	Error Trace:	/home/pete/Documents/_cake/kin-openapi.git/openapi3filter/issue201_test.go:134
     |             	Error:      	Received unexpected error:
     |             	            	response header "X-Blap" doesn't match schema: Value is not nullable
     |             	            	Schema:
     |             	            	  {
     |             	            	    "pattern": "^blap$"
     |             	            	  }
     |             	            	
     |             	            	Value:
     |             	            	  null
     |             	Test:       	TestIssue201/missing_non-required_header
```

**Comment by slessard** [comment]
Sorry for all the back and forth on this PR. All checks now pass.

**Review by fenollp — APPROVED** [review]

**Pull request merged**